Use structural pattern matching to handle linkcheck result statuses

jayaddison opened the pull request:
### Feature or Bugfix
- Refactoring (and a legacy code cleanup)

### Purpose
- The repeated `[el]if...result.status == <status>` conditions in part of the `linkcheck` code expose an intended behaviour: each status code implies that the code should follow one particular code path to handle the result.  This seems like a good fit for a [`match` clause](https://peps.python.org/pep-0622/), available from Python3.11+.

### Detail
- Handle `linkcheck` status codes using a `match` statement.
- Also includes a tangential refactor of some logging code, to fit the resulting code within formatting guidelines while remaining aesthetically reasonable.
- For review purposes, it could be helpful to ignore whitespace differences introduced by this changeset (`w=1` query-string parameter, if using the GitHub web comparison interface).

### Relates
- Builds upon #13043.

jayaddison committed:
Refactor: use Structural Pattern Matching (`match`) to handle linkstatus codes

jayaddison committed:
linkcheck: refactor logging code

This anticipates an additional level of indendation to be added for
the relevant code during the introduction of Structural Pattern
Matching.

AA-Turner reviewed (commented):
Can we integrate the `UNCHECKED` and `WORKING` (old) checks above into the `match`? Ideally, as we use an enum, we shouldn't need to use `case _` -- we just check every valid state.

A

jayaddison committed:
linkcheck: remove unused `old` message tag

Relates-to commit 43517219b7813c0c48d8a0c69656a8de7ab1df91.

URLs are de-duplicated by ``CheckExternalLinksBuilder.hyperlinks``.

jayaddison committed:
linkcheck: exhaustively check status codes

jayaddison committed:
Lint fixup: apply `ruff format`

jayaddison commented on sphinx/builders/linkcheck.py:
```diff
@@ -111,99 +111,97 @@ def process_result(self, result: CheckResult) -> None:
 
-        if result.status == _Status.UNCHECKED:
-            return
-        if result.status == _Status.WORKING and result.message == 'old':
-            return
```
Commenting for future historians: the `old` message tag here was a way that the `linkcheck` component would communicate to itself that a hyperlink had already been checked and confirmed as working ('good').

That code was removed in 43517219b7813c0c48d8a0c69656a8de7ab1df91, and I believe that the URI deduplication that was removed there is no longer necessary, because `linkcheck` nowadays collects hyperlinks in a Python `dict`, `CheckExternalLinksBuilder.hyperlinks`, keyed by the URI; so duplicates should not occur anyway.

AA-Turner committed:
Collapse if
